Initial charm import

seb128 opened the pull request:
Initial charm version, seems to work fine locally and I would welcome some feedback. It is currently missing tests and could use systemd timers instead of a crontab but those shouldn't be blocker for landing a first version in staging.

seb128 committed:
Initial charm import

jnsgruk commented on src/.charm.py.swp:
Oopsie :)

seb128 replied:
Removed

jnsgruk commented on src/charm.py:
```diff
@@ -0,0 +1,139 @@
+        except KeyError:
+            logger.warning("No 'gpg-secret-id' config, can't set up signing key")
+            self.unit.status = ops.ActiveStatus(
+                "No 'gpg-secret-id' config, signing will not be available."
```
```suggestion
                "Signing disabled. Set the 'gpg-secret-id' to enable."
```

jnsgruk commented on src/charm.py:
```diff
@@ -0,0 +1,139 @@
+        base = event.params["base"]
+
+        self.unit.status = ops.MaintenanceStatus(
+            f"Building new langpacks, base={base}, release={release}"
```
I'd maybe shorten this to "Building langpacks. See `juju debug-log`"

We want status messages to be short and sharp, giving a clear indication of required action.

Same applies elsewhere.

jnsgruk commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+            logger.debug("Crontab configured.")
+            return
+        except CalledProcessError as e:
+            logger.debug(f"Installation of the crontab failed: {e.stdout}")
```
```suggestion
            logger.debug(
            "Installation of the crontab failed: '%s'", e.stdout)
```

In general we try to avoid f-strings in logger calls

jnsgruk replied:
(Applies elsewhere)

jnsgruk commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+                apt.add_package(p)
+                logger.debug(f"Package {p} installed")
+            except PackageNotFoundError:
+                logger.error(f"failed to find package {p} in package cache")
```
Nit: some inconsistency in starting log strings with capitals or not throughout this file

seb128 replied:
Resolved in the new commits

jnsgruk commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+        """Update the langpack-o-matic checkout."""
+        # Pull Vcs updates
+        try:
+            run(
+                [
+                    "sudo",
+                    "-u",
+                    "ubuntu",
+                    "git",
+                    "-C",
+                    REPO_LOCATION,
+                    "pull",
+                ],
+                check=True,
+                stdout=PIPE,
+                stderr=STDOUT,
+                text=True,
+            )
```
This pattern of calling `run` with `sudo -u ubuntu` and catching the `CalledProcessError` is common enough that some sort of helper might make the code a little more readable, but not mandatory to progress.

seb128 replied:
I moved away from changing user for the job which resolves that point

jnsgruk commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+
+    def build_langpacks(self, base, release):
+        """Build the langpacks."""
+        lp = Launchpad.login_anonymously("langpacks", "production")
```
You may want to dependency-inject an LP client you own here to make testing this class easier (similar to https://github.com/canonical/ubuntu-manpages-operator/blob/main/src/launchpad.py)

jnsgruk commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+
+        # Download the current translations tarball from launchpad
+        try:
+            run(
+                [
+                    "sudo",
+                    "-u",
+                    "ubuntu",
+                    "wget",
+                    "--no-check-certificate",
+                    "-q",
+                    "-O",
+                    tarball,
+                    download_url,
+                ],
+                check=True,
+                stdout=PIPE,
+                stderr=STDOUT,
+                text=True,
+            )
```
Why call out to `wget` here and not use something like `requests`? (The latter will be much easier to test)

seb128 replied:
Because the python code is running as root and the service/data is owned by the "ubuntu" user. I don't know if there is a better way to handle that in the charm...
(it would be easier if the service was running into the root account the container but it is building deb packages and the Debian packaging tools strongly encourage to do builds as non privileged users)

seb128 replied:
Resolved, not trying to run as the ubuntu user allowed to use requests directly

jnsgruk commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+                    "sudo",
+                    "-u",
+                    "ubuntu",
+                    "git",
```
There are a few instances of invoking Git here, I wonder if in the long term adopting something like `GitPython` might be a better option?

seb128 replied:
Similarly to the wget comment, the charm is currently design to have the checkout owned by the "ubuntu" user...

jnsgruk reviewed (commented):
A few comments, but nothing blocking. Overall this a great start, modulo needing some tests, of course :)

matthew-hagemann commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+                    "clone",
+                    "-b",
+                    "master",
+                    "https://git.launchpad.net/langpack-o-matic",
```
I ran into the git clone hanging for me for 12 minutes on one run through of testing the charm. Might be worth adding a timeout.

seb128 replied:
I've switched to GitPython now but that doesn't seem to provide a timeout option :-/ I will keep that as a backlog improvement

matthew-hagemann commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+            logger.debug(f"Build of bin/msgequal failed: {e.stdout}")
+            raise
+
+    def _clean_builddir(self, releasedir):
```
I'd add type hints for all function arguments just to make it easier to work with them in `charm.py`

seb128 replied:
Thanks for the suggestion that is indeed nicer, resolved in the new commits

matthew-hagemann commented on src/langpacks.py:
```diff
@@ -0,0 +1,332 @@
+                text=True,
+            )
+            logger.debug(f"GPG key imported: {response.stdout}")
+        except CalledProcessError as e:
```
I ran into this error when adding a GPG key
`Jun 27 14:38:46 juju-19cb21-4 gpg-agent[6687]: command 'IMPORT_KEY' failed: Inappropriate ioctl for device <Pinentry>` 

I think this is due to me putting a pin on this gpg key. If that is the issue, we could have an optional pin be a separate secret that then can be fed into the command if present?

seb128 replied:
Thanks for the feedback, that's a good point. The issue isn't the pin but the fact that the setup expects a private key without a passphrase and hasn't a mechanism to prompt for the passphrase if there is one (the service is running as a cron job). Unsure if/how we could handle that better but I will try to a least document it.

jnsgruk replied:
If the key needs a passphrase, you could always add it as a secret to the charm's config

seb128 replied:
We could store the passphrase but then we would need logic to inject into the debsign prompt at the time the packages are built and signed which is extra work. The key is the one from the launchpad account used for uploads so a well defined one and it has been generated without passphrase so that isn't a problem currently. I'm going to keep that as a future improvement.

matthew-hagemann commented on src/charm.py:
```diff
@@ -0,0 +1,139 @@
+            return
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_upload_langpacks(self, event: ops.ActionEvent):
```
I don't know the `langpack-o-matic's` usage the gpg key, but I'm guessing you need it at least when uploading? If so, I'd add a check to make sure the secret is present to catch that issue early before attempting if no gpg key is set.

seb128 replied:
Better indeed. I've done that now by using a StoredState value which seems to work fine (the alternative would be to get by poking into the env / using the gpg command which felt are bit more work / less robust)

jnsgruk replied:
As above - sadly StoredState has a few footguns and you're better off implementing a simple method to check if there is a key

matthew-hagemann reviewed (commented):
Looking really good!

I've added a few comments from some issues I spotted when local testing, but general structure and flow looks solid :slightly_smiling_face:

seb128 committed:
Update src/charm.py

Co-authored-by: Jon Seager <jon@sgrs.uk>

seb128 committed:
Update src/langpacks.py

Co-authored-by: Jon Seager <jon@sgrs.uk>

seb128 committed:
Remove a .swp commited by error

seb128 committed:
Don't use f-string in logger calls

seb128 committed:
Start strings with a capital

seb128 committed:
Tweak the status messages to be more direct

seb128 committed:
Catch also ModelError

secret_get can raise "ops.model.ModelError: ERROR permission denied"

seb128 committed:
Include type hints to the functions arguments

seb128 committed:
Don't try to upload if there is no signing key configured

seb128 committed:
Formatting

seb128 committed:
Download the tarball using requests

seb128 committed:
Log a warning if a signing key is missing for upload

seb128 committed:
Don't use BlockedStatus for cases where the service is still working.

seb128 committed:
Use GitPython bindings

seb128 committed:
Try to run without the commands without "sudo -u ubuntu"

seb128 committed:
Run make format

seb128 committed:
Extra tweaks to avoid userid changes

seb128 committed:
Use a StoredState for the has_signing_key variable

seb128 committed:
Define a new LaunchpadClient class to make testing easier

seb128 commented:
Thanks for the review feedback. I've reworked a bit the code to address the feedback. I also change the approach to not start the actions as "ubuntu" user which simplifies things. dpkg-buildpackage doesn't need privileges but is working fine as root as that reduces friction and shouldn't be a problem in a VM anyway

jnsgruk commented on src/charm.py:
```diff
@@ -0,0 +1,153 @@
+
+        self._langpacks = Langpacks(LaunchpadClient())
+
+        self._stored.set_default(has_signing_key=False)
```
In general I would avoid using StoredState completely - you're better off just checking on disk / checking the secret to verify if the charm has a key

seb128 replied:
ack, I've changed that to call gpg --list-secret-keys instead

seb128 committed:
Check for the signing key directly on disk and don't store the status

seb128 committed:
use https proxy for lplib when set

This is a localised fix which passes Launchpad.login_anonymously's
proxy_info keyword argument a callable which can elicit the correct
Juju proxy configuration from the model for both http and https schemes.

This is inspired by the upstream httplib2 proxy_info_from_environment
method which is used in the launchpadlib implementation.

seb128 committed:
Refresh the dependencies

seb128 commented:
Thanks for the review feedback. I've included an extra commit to handle proxy for launchpadlib and I'm going to merge that first PR now then iterate for tests and other improvements with new PRs
